lectures/hopenhayn.md

```diff
@@ -0,0 +1,871 @@
+
+Below we solve for the zero of this function to calculate $p*$.
+
+From the figure it looks like $p^*$ will be close to 1.5.
+
+
+
+### Computing the equilibrium
+
+
+
+Now let's try computing the equilibrium
+
+```{code-cell} ipython3
+%%time
+p_star, v_bar, φ_star, φ_sample, s, m_star = \
+        compute_equilibrium_prices_and_quantities(model)
+```
+
+Let's check that $p^*$ is close to 1.5
+
+```{code-cell} ipython3
+p_star
+```
+
+Here is a plot of the value function $\bar v(\cdot, p^*)$.
+
+```{code-cell} ipython3
+fig, ax = plt.subplots()
+ax.plot(grids.φ_grid, v_bar, label=r'$\varphi \mapsto \bar v(\varphi, p^*)$')
+ax.set_xlabel("productivity")
+ax.set_ylabel("firm value")
+ax.legend()
+plt.show()
+```
+
+Let's have a look at the firm size distribution, with firm size measured by output.
+
+```{code-cell} ipython3
+output_dist = q(φ_sample, p_star, parameters)
+```
+
+```{code-cell} ipython3
+fig, ax = plt.subplots(figsize=(7, 4))
+ax.hist(jnp.log(output_dist), bins=100, density=True,
+        label="firm size distribution")
+ax.set_xlabel("log output")
+ax.set_ylabel("frequency")
+ax.legend()
+plt.show()
+```
+
+## Pareto tails
+
+
+
+The firm size distribution shown above appears to have a long right tail.
+
+This matches the observed firm size distribution.
+
+In fact the firm size distribution obeys a **power law**.
+
+More mathematically, the distribution of firm size has a Pareto right hand tail, so that there exist constants $k, \alpha > 0$ with
+
+$$
+   \mu((x, \infty)) \approx kx^{-\alpha} \text{ when $x$ is large}
+$$
+
+Here $\alpha$ is called the tail index.
+
+Does the model replicate this feature?
+
+
+
+One option is to look at the empirical counter CDF (cumulative distribution).
+
+The idea is as follows: The counter CDF of a random variable $X$ is 
+
+$$
+   G(x) := \mathbb P\{X > x\}
+$$
+
+In the case of a Pareto tailed distribution we have $\mathbb P\{X > x\} \approx k x^{-\alpha}$ for large $x$.
+
+Hence, for large $x$,
+
+$$
+    \ln G(x) \approx \ln k - \alpha \ln x
+$$
+
+The empirical counterpart of $G$ given sample $X_1, \ldots, X_n$ is 
+
+$$
+    G_n (x) := \frac{1}{n} \sum_{i=1}^n \mathbf 1 \{X_i > x \}
+$$
+
+For large $k$ (implying $G_n \approx G$) and large $x$, we expect that, for a Pareto-tailed sample, $\ln G_n$ is approximately linear.
+
+
+
+Here's a function to compute the empirical counter CDF:
+
+```{code-cell} ipython3
+def ECCDF(data):
+    """
+    Return a function that implements the ECCDF given the data.
+    """
+    def eccdf(x):
+        return jnp.mean(data > x)
+    return eccdf
+```
+
+Let's plot the empirical counter CDF of the output distribution.
+
+```{code-cell} ipython3
+eccdf = ECCDF(output_dist)
+
+ϵ = 10e-10
+x_grid = jnp.linspace(output_dist.min() + ϵ, output_dist.max() - ϵ, 200)
+y = [eccdf(x) for x in x_grid]
+
+fix, ax = plt.subplots()
+ax.loglog(x_grid, y, 'o', label="ECCDF")
+ax.set_xlabel("productivity")
+ax.set_ylabel("counter CDF")
+plt.show()
+```
+
+**Exercise**
```

Comment:
Please put this in an exercise environment, and the solution below in a solution environment.